### PR TITLE
Feat/56

### LINKED ISSUE #56: Create HardHat project to compile and test the `RenPool` contract
With this issue, we start the migration to _HardHat_, from _Brownie_.

The _HardHat_ project should be able to compile the `RenPool` contract properly. This means, it should be configured with a suitable compiler and _OpenZeppelin_ dependencies.

The _HardHat_ project should use `yarn` to manage its dependencies.

### PR DESCRIPTION
**Pull request by acuarica**
Closes #56.

### CODE REVIEW

**Review by fede-rodes — COMMENTED** [review]
Just some comments to get started!

File: contracts/RenPool.sol [hunk 1]
```diff
@@ -132,17 +133,14 @@ contract RenPool {
...
+		require(_amount + totalPooled <= bond, "RenPool: Amount surpasses bond");
 
 		balances[sender] += _amount;
 		totalPooled += _amount;
 
-		require(
-			renToken.transferFrom(sender, address(this), _amount) == true,
-			"Deposit failed"
-		);
+		renToken.transferFrom(sender, address(this), _amount);
```
[the remark this answers is not in the document]
Comment (acuarica) [reply contracts/RenPool.sol:144]: The signature of `transferFrom` indicates that it returns `bool`, so the `require` seems indeed correct. The problem is that the implementation used by `RenToken` of `ERC20`[1] is

```solidity
    function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
        _transfer(sender, recipient, amount);
        _approve(sender, _msgSender(), _allowances[sender][_msgSender()].sub(amount, "ERC20: transfer amount exceeds allowance"));
        return true;
    }
```

which always returns `true`. Thus, the `require` is misleading, that's the reason I've removed it.

---

[1] https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/release-v2.x-erc1155/contracts/token/ERC20/ERC20.sol#L100

File: test/RenPool.test.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  ethers: {
+    utils: { base58 },
+    BigNumber: { from: bn }
+  },
+  network: {
+    config: {
+      renTokenAddr,
+      renBTC,
+      topRenTokenHolder,
+      darknodeRegistryAddr,
```
[the remark this answers is not in the document]
Comment (acuarica) [reply test/RenPool.test.js:15]: Agree, will fix it.

File: test/RenPool.test.js [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    for (const user of (await ethers.getSigners()).slice(2)) {
+      expect(await renToken.connect(owner).balanceOf(user.address)).to.equal(0);
+
+      const amount = POOL_BOND.mul(5);
+      await renToken.connect(signer).transfer(user.address, amount);
+      expect(await renToken.connect(owner).balanceOf(user.address)).to.equal(amount);
+    }
+
+    await provider.request({ method: 'hardhat_stopImpersonatingAccount', params: [topRenTokenHolder] });
+  });
```
[the remark this answers is not in the document]
Comment (acuarica) [reply test/RenPool.test.js:59]: Yes, we can do it like this as well. The main reason I did it on the `before` hook is that the state is shared for all tests. There are two test methods provided by Hardhat, `evm_snapshot` and `evm_revert` that can be used to control the state of the test blockchain. With these methods, we can implement it as you mention.

[the remark this answers is not in the document]
Comment (fede-rodes) [reply test/RenPool.test.js:59]: Ok, got it :+1:

**Review by fede-rodes — COMMENTED** [review]
2 more comments

File: test/RenPool.test.js [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    expect(await renPool.bond()).to.equal(POOL_BOND);
+    expect(await renPool.isLocked()).to.equal(false);
+    expect(await renPool.totalPooled()).to.equal(0);
+  });
+
+  describe('deposit', function () {
+
+    it('should deposit REN into RenPool', async function () {
+      const amount = 1;
+
```
[the remark this answers is not in the document]
Comment (acuarica) [reply test/RenPool.test.js:95]: Yes, totally agree.

File: test/RenPool.test.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+      await renToken.connect(alice).approve(renPool.address, POOL_BOND);
+      await renPool.connect(alice).fulfillWithdrawRequest(bob.address);
+
+      expect(await renToken.connect(owner).balanceOf(renPool.address)).to.equal(POOL_BOND);
+      expect(await renToken.connect(alice).balanceOf(alice.address)).to.equal(aliceBalance.sub(amount));
+      // expect(await renToken.connect(bob).balanceOf(bob.address)).to.equal(bobBalance);
+      expect(await renPool.balanceOf(alice.address)).to.equal(amount);
+      // expect(await renPool.balanceOf(bob.address)).to.equal(0);
+      expect(await renPool.isLocked()).to.be.true;
```
Comment (fede-rodes) [inline test/RenPool.test.js:226]: Good catch! Why bob's balance is not being updated properly?

Comment (fede-rodes) [reply test/RenPool.test.js:226]: Ok, solidity code seems fine, there was some issues with the tests. I think it should the as follows:

```
      expect(await renToken.connect(bob).balanceOf(bob.address)).to.equal(bobBalance.sub(POOL_BOND).add(amount));
      expect(await renPool.balanceOf(bob.address)).to.equal(POOL_BOND.sub(amount));
```

File: package.json [hunk 6]
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "renpool",
+  "version": "0.1.0",
+  "repository": "git@github.com:Ethernautas/renpool.git",
+  "author": "Fede Rodes",
+  "license": "MIT",
+  "scripts": {
+    "build": "hardhat compile",
+    "test": "hardhat test",
+    "test:mainnet": "FORK=mainnet hardhat test",
```
Comment (acuarica) [inline package.json:10]: @fede-rodes here's how I solve the config issue. The `FORK` env var indicates where to fork from when running tests.

Check `hardhat.config.js` for details.

Comment (fede-rodes) [reply package.json:10]: Perfect!

**Review by fede-rodes — APPROVED** [review]

**Pull request merged**